### ISSUE TITLE
Client throws "empty response" exception for all curl errors

### ISSUE THREAD

**Issue by rossmotley** [sailthru/sailthru-php5-client#71]
Hi,

When debugging a timeout from the Sailthru API, I noticed that currently the code is throwing an "empty response" exception for all curl errors:

```php
        $response = curl_exec($ch);
        // ...

        if (!$response) {
            throw new Sailthru_Client_Exception(
                "Bad response received from $url",
                Sailthru_Client_Exception::CODE_RESPONSE_EMPTY
            );
        }
```

Here, `$response` is the response body or `false` to indicate an error from curl. It's easy to get the actual error with `curl_error()`, which could be added as a new exception type, something like: `Sailthru_Client_Exception::CODE_HTTP_ERROR` or similar. That would provide a much more helpful error for developers.

I'm happy to make a little PR for this, but wanted to check that it would be accepted, as it's a BC break of some sort.

Thanks,

Ross

**Comment by silverman63** [comment]
Hi Ross, thanks for submitting this. Totally makes sense.

What I'd like to do is incorporate this into a 2.0.0 release, as throwing more exceptions was already a bit in the direction of breaking changes. Feel free to make the PR, and we can discuss implementation for 2.0.0.

**Comment by rossmotley** [comment]
Ok - the other thing I was thinking was that it would be great to extract the logic currently in `\Sailthru_Client::httpRequest()` into a separate interface/classes. 

This would make it possible to test the `Sailthru_Client` in isolation etc. That would probably make sense to do at the same time as changing the exceptions.

**Comment by rossmotley** [comment]
@silverman63 Totally untested, but here is a sketch of what I was thinking:

https://github.com/sailthru/sailthru-php5-client/compare/master...rossmotley:transport-extract?expand=1

**Comment by silverman63** [comment]
@rossmotley I'll have to explore this further, but my instinct is to keep these separate for prioritization and time's sake. Will get back to you.

**Comment by silverman63** [comment]
@rossmotley did you ever make a PR for this? Additionally, can you give me steps to reproduce the curl error?

**Comment by rossmotley** [comment]
@silverman63 I have created #78 without the separation of the transport into its own class.

**Comment by rossmotley** [comment]
Implemented with #79
